core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java

```diff
@@ -0,0 +1,192 @@
+            }
+            properties = Collections.unmodifiableMap(props);
+        }
+
+        @Override
+        public Map<String, ? extends PojoProperty<T>> getProperties() {
+            return properties;
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public PojoBuilder<T> create() {
+            final Class<?> type = GenericTypes.getErasedType(getType());
+            final T instance;
+            try {
+                instance = (T) type.newInstance();
+            } catch (Exception e) {
+                throw new IllegalArgumentException(String.format(TYPE_NOT_INSTANTIABLE, type.getName()), e);
+            }
+            return new PojoBuilder<T>() {
+                @Override
+                public void set(String property, Object value) {
+                    final BeanPojoProperties<T>.BeanPojoProperty<T> prop = properties.get(property);
+                    try {
+                        Method writeMethod = prop.descriptor.getWriteMethod();
+                        if (writeMethod == null) {
+                            throw new IllegalArgumentException(String.format(MISSING_SETTER, property));
+                        }
+                        writeMethod.invoke(instance, value);
+                    } catch (IllegalAccessException e) {
+                        throw new IllegalArgumentException(String.format(SETTER_NOT_ACCESSIBLE, property), e);
+                    } catch (InvocationTargetException e) {
+                        throw new IllegalArgumentException(String.format(INVOCATION_TARGET_EXCEPTION, property), e);
+                    } catch (IllegalArgumentException e) {
+                        throw new IllegalArgumentException(String.format(REFLECTION_ILLEGAL_ARGUMENT_EXCEPTION,
+                            prop.getQualifiedType(), prop.getName()), e);
+                    }
+                }
+
+                @Override
+                public T build() {
+                    return instance;
+                }
+            };
+        }
+
+        class BeanPojoProperty<T> implements PojoProperty<T> {
+            final PropertyDescriptor descriptor;
+
+            BeanPojoProperty(PropertyDescriptor property) {
+                this.descriptor = property;
+            }
+
+            @Override
+            public String getName() {
+                return descriptor.getName();
+            }
+
+            @Override
+            public QualifiedType<?> getQualifiedType() {
+                Parameter setterParam = Optional.ofNullable(descriptor.getWriteMethod())
+                    .map(m -> m.getParameterCount() > 0 ? m.getParameters()[0] : null)
+                    .orElse(null);
+
+                return QualifiedType.of(
+                    Optional.ofNullable(descriptor.getReadMethod())
+                        .map(Method::getGenericReturnType)
+                        .orElseGet(() -> descriptor.getWriteMethod().getGenericParameterTypes()[0]))
```

Comment:
I'm not sure if we were doing this before, but should we resolve these types against the bean class? The properties might parameterized from a supertype.

Reply:
We weren't doing that before.  I agree with you it would make sense to do so, but can we save that for another PR?  I'm trying to match existing bean infrastructure behavior as closely as possible to reduce risk of this PR.